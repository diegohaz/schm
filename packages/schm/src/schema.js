// @flow
import merge from 'lodash/merge'
import type { Schema, SchemaGroup } from './types'
import { isSchema } from './utils'
import parsers from './parsers'
import validators from './validators'
import parse from './parse'
import validate from './validate'
import parseParams from './parseParams'

export const defaultSchema = (params?: Object = {}): Schema => ({
  parsers,
  validators,
  parse(values) {
    return parse(values, this)
  },
  validate(values, paramPathPrefix) {
    return validate(values, this, paramPathPrefix)
  },
  merge(...schemas) {
    const merged = merge({}, this, ...schemas)
    return {
      ...merged,
      params: parseParams(merged.params),
    }
  },
  params: parseParams(params),
})

/**
 * A simple group of parameters. It's used internally when you pass literal
 * objects to [`schema`](#schema).
 * @example
 * const userSchema = schema(
 *   group({
 *     id: String,
 *     name: String,
 *   }),
 *   group({
 *     age: Number,
 *   })
 * )
 */
export const group = (params?: Object = {}): SchemaGroup => (
  (previous: Schema): Schema => previous.merge(defaultSchema(params))
)

/**
 * Creates a schema by composing groups of parameters.
 * @example
 * const userSchema = schema({
 *   id: String,
 *   name: String,
 * }, {
 *   age: Number
 * })
 *
 * // nested schema
 * const teamSchema = schema({
 *   users: [userSchema],
 * })
 */
const schema = (...groups: (Object | Schema | SchemaGroup)[]): Schema =>
  groups.reduce((finalSchema, currentGroup) => {
    if (typeof currentGroup === 'function') {
      return currentGroup(finalSchema)
    } else if (isSchema(currentGroup)) {
      return currentGroup
    }
    return group(currentGroup)(finalSchema)
  }, defaultSchema())

export default schema
