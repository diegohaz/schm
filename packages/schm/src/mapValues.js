// @flow
import { toArray, isArray, isSchema } from "./utils";

type TransformValueFunction = (
  value: any,
  options: Object,
  paramName: string,
  paramPath: string
) => any;

type MapValuesFunction = (
  values: Object,
  params: Object,
  transformValueFn: TransformValueFunction,
  paramNames?: any[]
) => Object;

const mapValues: MapValuesFunction = (
  values,
  params,
  transformValueFn,
  paramNames = []
) =>
  Object.keys(params).reduce((finalParams, paramName) => {
    const options = params[paramName];
    const value = values[paramName];
    const paramPath = [...paramNames, paramName].join(".");
    const mergeParam = finalValue => ({
      ...finalParams,
      [paramName]: finalValue
    });

    if (isArray(options.type)) {
      const [opt] = options.type;
      const finalValue = toArray(value).map((val, i) =>
        transformValueFn(val, opt, paramName, [paramPath, i].join("."))
      );
      return mergeParam(finalValue);
    }

    if (isSchema(options.type)) {
      return mergeParam(
        mapValues(value, options.type.params, transformValueFn, [paramPath])
      );
    }

    return mergeParam(transformValueFn(value, options, paramName, paramPath));
  }, {});

export default mapValues;
