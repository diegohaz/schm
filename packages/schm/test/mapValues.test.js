import faker, { name, lorem } from 'faker'
import times from 'lodash/times'
import mapValues from '../src/mapValues'
import schema from '../src/schema'

const createPerson = () => ({ name: name.firstName(), age: 20 })
const createStudent = () => ({ ...createPerson(), grade: 5 })
const createTeacher = () => ({ ...createPerson(), subjects: times(3, lorem.word) })
const createClass = () => ({
  grade: 5,
  subject: lorem.word(),
  teacher: createTeacher(),
  students: times(5, createStudent),
})

const personSchema = schema({
  name: String,
  age: Number,
})
const studentSchema = schema(personSchema, {
  grade: Number,
})
const teacherSchema = schema(personSchema, {
  subjects: [String],
})
const classSchema = schema({
  grade: Number,
  subject: String,
  teacher: teacherSchema,
  students: [studentSchema],
})

beforeEach(() => {
  faker.seed(123)
})

it('maps values correctly', () => {
  const values = createClass()
  const transformValue = value => value
  expect(mapValues(values, classSchema.params, transformValue))
    .toMatchSnapshot()
})
