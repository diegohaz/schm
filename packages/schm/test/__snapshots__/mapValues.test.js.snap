// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`maps values correctly 1`] = `
Object {
  "grade": 5,
  "students": Array [
    Object {
      "age": 20,
      "grade": 5,
      "name": "Maryse",
    },
    Object {
      "age": 20,
      "grade": 5,
      "name": "Kathryne",
    },
    Object {
      "age": 20,
      "grade": 5,
      "name": "Miller",
    },
    Object {
      "age": 20,
      "grade": 5,
      "name": "Millie",
    },
    Object {
      "age": 20,
      "grade": 5,
      "name": "Jennings",
    },
  ],
  "subject": "optio",
  "teacher": Object {
    "age": 20,
    "name": "Mervin",
    "subjects": Array [
      "quis",
      "molestiae",
      "tempora",
    ],
  },
}
`;
